Enable automatic cross-thread stack tracing

extesy opened the pull request:
Normally tasks are executed in a different thread from the one creating it and at a different time. This makes it hard to debug because when a task throws an exception its call stack ends in the execution engine that actually starts a thread that serves the task. This feature collects task's stack trace in advance, when task is created, so that if the task throws the exception then the parent stack trace is appended to it. This has a small performance impact even if the task doesn't throw any exception because stack trace is collected in task constructor.

Stack trace before:

```
java.lang.RuntimeException: error
  at com.linkedin.parseq.TestTask.lambda$getFailureTask$2(TestTask.java:135)
  at com.linkedin.parseq.TestTask$$Lambda$91/209387180.call(Unknown Source)
  at com.linkedin.parseq.Task.lambda$async$20(Task.java:1060)
  at com.linkedin.parseq.Task$$Lambda$32/1381965390.apply(Unknown Source)
  at com.linkedin.parseq.Task$1.run(Task.java:1097)
  at com.linkedin.parseq.BaseTask.doContextRun(BaseTask.java:232)
  at com.linkedin.parseq.BaseTask.contextRun(BaseTask.java:200)
  at com.linkedin.parseq.internal.ContextImpl$2.run(ContextImpl.java:93)
  at com.linkedin.parseq.internal.SerialExecutor$ExecutorLoop.run(SerialExecutor.java:121)
  at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
  at java.util.concurrent.FutureTask.run(FutureTask.java:266)
  at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
  at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
  at java.lang.Thread.run(Thread.java:745)
```

Stack trace after:

```
java.lang.RuntimeException: error
  at com.linkedin.parseq.TestTask.lambda$getFailureTask$2(TestTask.java:135)
  at com.linkedin.parseq.TestTask$$Lambda$91/833726670.call(Unknown Source)
  at com.linkedin.parseq.Task.lambda$async$20(Task.java:1060)
  at com.linkedin.parseq.Task$$Lambda$32/1381965390.apply(Unknown Source)
  at com.linkedin.parseq.Task$1.run(Task.java:1097)
  at ********** Task "failure" (above) was instantiated as following (below): **********.(Unknown Source)
  at com.linkedin.parseq.Task$1.<init>(Task.java:1094)
  at com.linkedin.parseq.Task.async(Task.java:1094)
  at com.linkedin.parseq.Task.async(Task.java:1058)
  at com.linkedin.parseq.TestTask.getFailureTask(TestTask.java:134)
  at com.linkedin.parseq.TestTask.lambda$testStackFrames$0(TestTask.java:116)
  at com.linkedin.parseq.TestTask$$Lambda$95/1205406622.apply(Unknown Source)
  at com.linkedin.parseq.promise.PromiseTransformer.accept(PromiseTransformer.java:34)
  at com.linkedin.parseq.promise.PromiseTransformer.accept(PromiseTransformer.java:6)
  at com.linkedin.parseq.FusionTask.lambda$adaptToAcceptTraceContext$0(FusionTask.java:73)
  at com.linkedin.parseq.FusionTask$$Lambda$6/878274034.accept(Unknown Source)
  at com.linkedin.parseq.FusionTask.lambda$completing$3(FusionTask.java:175)
  at com.linkedin.parseq.FusionTask$$Lambda$7/1296674576.accept(Unknown Source)
  at com.linkedin.parseq.FusionTask.propagate(FusionTask.java:321)
  at com.linkedin.parseq.FusionTask.lambda$run$6(FusionTask.java:342)
  at com.linkedin.parseq.FusionTask$$Lambda$21/2049720174.apply(Unknown Source)
  at com.linkedin.parseq.Task$1.run(Task.java:1097)
  at ********** Task "nested" (above) was instantiated as following (below): **********.(Unknown Source)
  at com.linkedin.parseq.Task$1.<init>(Task.java:1094)
  at com.linkedin.parseq.Task.async(Task.java:1094)
  at com.linkedin.parseq.Task.flatten(Task.java:829)
  at com.linkedin.parseq.Task.flatMap(Task.java:232)
  at com.linkedin.parseq.TestTask.testStackFrames(TestTask.java:116)
  at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
  at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  at java.lang.reflect.Method.invoke(Method.java:497)
  at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:85)
  at org.testng.internal.Invoker.invokeMethod(Invoker.java:639)
  at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:816)
  at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1124)
  at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:125)
  at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
  at org.testng.TestRunner.privateRun(TestRunner.java:774)
  at org.testng.TestRunner.run(TestRunner.java:624)
  at org.testng.SuiteRunner.runTest(SuiteRunner.java:359)
  at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:354)
  at org.testng.SuiteRunner.privateRun(SuiteRunner.java:312)
  at org.testng.SuiteRunner.run(SuiteRunner.java:261)
  at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
  at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
  at org.testng.TestNG.runSuitesSequentially(TestNG.java:1215)
  at org.testng.TestNG.runSuitesLocally(TestNG.java:1140)
  at org.testng.TestNG.run(TestNG.java:1048)
  at org.testng.TestNG.privateMain(TestNG.java:1355)
  at org.testng.TestNG.main(TestNG.java:1324)
```

jodzga commented:
Could you also put an example of a stack trace before and after this change?

jodzga commented on src/com/linkedin/parseq/BaseTask.java:
```diff
@@ -122,7 +125,13 @@ public BaseTask(final String name, final String taskType) {
-    _stateRef = new AtomicReference<State>(state);
+    _stateRef = new AtomicReference<>(state);
+
+    if (ParSeqGlobalConfiguration.getInstance().isCrossThreadStackTracesEnabled()) {
```
ParSeqGlobalConfiguration.getInstance().isCrossThreadStackTracesEnabled() looks a bit verbose. I think that maybe having static methods in ParSeqGlobalConfiguration directly is a better idea after all.

extesy replied:
Sounds good

jodzga commented on src/com/linkedin/parseq/BaseTask.java:
```diff
@@ -311,12 +320,71 @@ private void done(final T value, final TaskLogger taskLogger) {
+  // Concatenate stack traces if kept the original stack trace from the task creation
+  private void appendTaskStackTrace(final Throwable error) {
+    if (!ParSeqGlobalConfiguration.getInstance().isCrossThreadStackTracesEnabled() || error == null ||
+        _taskStackTrace == null || _taskStackTrace.length <= 2) {
```
What is the case when _taskStackTrace == 2? Can you add comment about it?

extesy replied:
Done

extesy commented:
You mean here in PR description or in the actual code somewhere?

jodzga commented on src/com/linkedin/parseq/BaseTask.java:
```diff
@@ -311,12 +320,71 @@ private void done(final T value, final TaskLogger taskLogger) {
+    }
+
+    // Skip stack frames up to the BaseTask (useless Thread.getStackTrace stuff)
+    int skipTaskFrames = 1;
```
Seems like a copy-paste. Can you refactor skipping logic to a separate method?

extesy replied:
There's not much actual copy-paste - these two loops go into opposite directions and compute index differently. If I try to combine both logics in a single function it will look a lot more confusing and it will still contain if/else to support bi-directional scan. I would recommend against it.

jodzga replied:
Got it. Didn't notice that they go opposite direction.

jodzga commented on test/com/linkedin/parseq/TestTask.java:
```diff
@@ -1,16 +1,34 @@
  */
 public class TestTask extends AbstractTaskTest {
 
+  private boolean _crossThreadStackTracesEnabled;
```
You can use DataProvider to have a better coverage and run all tests with and without this option enabled.
I think you can also use DataProvider with the @BeforeXXX methods.

extesy replied:
Unfortunately DataProvider can't be used with @BeforeXXX methods so running all tests with/without this option would require modifying each indivual test to accept and process this parameter.

jodzga replied:
Actually it is supported but not well documented, see e.g.:
https://github.com/jodzga/parseq/blob/perf-improvements/test/com/linkedin/parseq/internal/TestSerialExecutor.java#L34

You just need to pass an Object[] to the method.

extesy replied:
Does it only work for BeforeMethod? Because for BeforeClass it passes null as `testArgs` instead of a flag.

jodzga replied:
That I don't know. I only used it with BeforeMethod. Unfortunately I did not find an official documentation about this.

extesy replied:
Well, it didn't work for me. I copy/pasted from your `TestSerialExectutor.java` example and it didn't work for BeforeClass.

jodzga commented on src/com/linkedin/parseq/ParSeqGlobalConfiguration.java:
```diff
@@ -0,0 +1,50 @@
+   * Returns current state of cross-thread (cross-task) stack tracing.
+   *
+   * Normally tasks are executed in different thread from the one creating it and at a different time. This makes it
+   * hard to debug because if task throws an exception, its call stack ends in the execution engine that actually
```
its -> it's

extesy replied:
I think it should be actually "its" here. If I replace with it's then the phrase "it is call stack ends" is not grammatically correct.

jodzga replied:
Yup, you are right :-)

jodzga commented on src/com/linkedin/parseq/ParSeqGlobalConfiguration.java:
```diff
@@ -0,0 +1,50 @@
+   * Modifies the current state of cross-thread (cross-task) stack tracing.
+   *
+   * Normally tasks are executed in different thread from the one creating it and at a different time. This makes it
+   * hard to debug because if task throws an exception, its call stack ends in the execution engine that actually
```
its -> it's

extesy replied:
`it's` is a shorthand for `it is` and that doesn't really fit into the phrase.

jodzga replied:
Yup.

jodzga commented on src/com/linkedin/parseq/ParSeqGlobalConfiguration.java:
```diff
@@ -0,0 +1,50 @@
+   * starts a thread that serves the task. This feature collects stack trace in advance, when task is created, so that
+   * if the task throws the exception then the parent stack trace is appended to it. This has a small performance
+   * impact even if the task doesn't throw any exception because stack trace is collected in task constructor.
+   *
```
Also add info that this is dynamic configuration and changing this setting at runtime will take effect immediately - this might not be obvious.

extesy replied:
Done

jodzga commented:
I meant PR description, thanks.

extesy commented on src/com/linkedin/parseq/ParSeqGlobalConfiguration.java:
```diff
@@ -0,0 +1,50 @@
+package com.linkedin.parseq;
```
sorry, forgot about it

extesy commented on src/com/linkedin/parseq/ParSeqGlobalConfiguration.java:
```diff
@@ -0,0 +1,50 @@
+
+/**
+ * Global parseq configuration, applies to all Engine and Task instances
+ */
```
done

extesy commented on src/com/linkedin/parseq/ParSeqGlobalConfiguration.java:
```diff
@@ -0,0 +1,50 @@
+public class ParSeqGlobalConfiguration {
+  private static final ParSeqGlobalConfiguration _instance = new ParSeqGlobalConfiguration();
+
+  private AtomicBoolean _crossThreadStackTracesEnabled = new AtomicBoolean(false);
```
Yep

jodzga commented:
Could you also update CHANGELOG?

jodzga commented:
I just merged a #113 PR to master, please rebase.

extesy commented:
Rebased and updated changelog. Somehow this pull request now also shows commits from the master, but there is no actual conflict. Do you want me to create a new PR instead?

jodzga commented on CHANGELOG.md:
```diff
@@ -1,3 +1,9 @@
+------
+
+* Deprecate Tasks.par(...) for safer alternative Task.par(...) that does not throw IllegalArgumentException on empty collection.
+* Enable automatic cross-thread stack tracing
```
+ "It is optional feature, turned off by default. See ParSeqGlobalConfiguration. setCrossThreadStackTracesEnabled()."

extesy replied:
Done
